alibi_detect/utils/tests/test_saving.py

```diff
@@ -1,242 +1,888 @@
+def nlp_embedding_and_tokenizer(model_name, max_len, uae, backend):
+    """
+    A fixture to build nlp embedding and tokenizer models based on the HuggingFace pre-trained models.
+    """
+    backend = 'tf' if backend == 'tensorflow' else 'pt'
+
+    # Load tokenizer
+    tokenizer = AutoTokenizer.from_pretrained(model_name)
+    X = 'A dummy string'  # this will be padded to max_len
+    tokens = tokenizer(list(X[:5]), pad_to_max_length=True,
+                       max_length=max_len, return_tensors=backend)
+
+    # Load embedding model
+    emb_type = 'hidden_state'
+    n_layers = 8
+    layers = [-_ for _ in range(1, n_layers + 1)]
+    enc_dim = 32
+
+    if backend == 'tf':
+        embedding = TransformerEmbedding_tf(model_name, emb_type, layers)
+        if uae:
+            x_emb = embedding(tokens)
+            shape = (x_emb.shape[1],)
+            embedding = UAE_tf(input_layer=embedding, shape=shape, enc_dim=enc_dim)
+    else:
+        embedding = TransformerEmbedding_pt(model_name, emb_type, layers)
+        if uae:
+            x_emb = embedding(tokens)
+            emb_dim = x_emb.shape[1]
+            device = torch.device(DEVICE)
+            embedding = torch.nn.Sequential(
+                embedding,
+                torch.nn.Linear(emb_dim, 256),
+                torch.nn.ReLU(),
+                torch.nn.Linear(256, enc_dim)
+            ).to(device).eval()
+
+    return tokenizer, embedding, max_len, enc_dim
+
+
+@fixture
+def preprocess_nlp(embedding, tokenizer, current_cases, max_len, backend):
+    """
+    Preprocess function with Untrained Autoencoder.
+    """
+    if backend == 'tensorflow':
+        preprocess_fn = partial(preprocess_drift_tf, model=embedding, tokenizer=tokenizer, max_len=max_len)
+    else:
+        preprocess_fn = partial(preprocess_drift_pt, model=embedding, tokenizer=tokenizer, max_len=max_len)
+    return preprocess_fn
+
+
+@fixture
+def preprocess_hiddenoutput(classifier, current_cases, backend):
+    """
+    Preprocess function to extract the softmax layer of a classifier (with the HiddenOutput utility function).
+    """
+    if backend == 'tensorflow':
+        model = HiddenOutput_tf(classifier, layer=-1)
+        preprocess_fn = partial(preprocess_drift_tf, model=model)
+    else:
+        model = HiddenOutput_pt(classifier, layer=-1)
+        preprocess_fn = partial(preprocess_drift_pt, model=model)
+    return preprocess_fn
+
+
+@parametrize('preprocess_fn', [preprocess_custom, preprocess_hiddenoutput])
+@parametrize_with_cases("data", cases=ContinuousData, prefix='data_')
+def test_save_ksdrift(data, preprocess_fn, tmp_path):
+    """
+    Test KSDrift on continuous datasets, with UAE and classifier softmax output as preprocess_fn's. Only this
+    detector is tested with preprocessing strategies, as other detectors should see the same preprocess_fn output.
+
+    Detector is saved and then loaded, with assertions checking that the reinstantiated detector is equivalent.
```

Comment:
This is definitely out of scope, but I wonder if it's useful/relevant at all to implement the Python dunder methods to test equality of specific detectors? Might be helpful in general for a user to distinguish quickly what is different between two detectors of the same type (essentially a diff between configs).

Reply:
This would be very nice. The current equivalence assertions are a real headache. I've added it to the issue list.